sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,703 @@
+// provided as uint32_t as fixed point numbers val / 2^32, so to compute
+// integer random in range [0..N) we can use multiplication instead of
+// finding remainder of division which is slow.
+//
+//          value = ((uint64_t)N * (uint64_t)seed) >> 32
+//
+// For uniformly distributed seed it gives us same practical result as
+//
+//          value = seed % N;
+//
+// Also, strides like +1, -1, +2 are trivial and not desirable, so we can
+// make strides starting +3 and higher. Adjustment of range (-4) eliminates
+// stride == -1 mod N == N-1 mod N.
+//
+// Compact representation of first prime numbers. Try to fit into
+// 8-bit, so store primes decremented by minimal prime number.
+// Prime numbers for reference:
+//  11     13     17     19     23     29    31     37     41     43
+//  47     53     59     61     67     71    73     79     83     89
+//  97    101    103    107    109    113   127    131    137    139
+// 149    151    157    163    167    173   179    181    191    193
+// 197    199    211    223    227    229   233    239    241    251
+// 257    263    269    271    277    281   283
+#define MIN_PRIME 11
+
+// Decrement MIN_PRIME from value.
+#define DMIN(p) (p - MIN_PRIME)
+
+// Maximal value accepted as input for larger_prime()
+#define PRIME_MAX 256
+
+// Return prime larger than N, when N is in range [0, 256]
+static uint32_t larger_prime(uint32_t n) {
+  static const uint8_t kPrime[] = {
+      DMIN(11) /* <=8 */,    DMIN(23) /* <=16 */,   DMIN(29) /* <=24 */,
+      DMIN(37) /* <=32 */,   DMIN(47) /* <=40 */,   DMIN(59) /* <=48 */,
+      DMIN(61) /* <=56 */,   DMIN(71) /* <=64 */,   DMIN(79) /* <=72 */,
+      DMIN(89) /* <=80 */,   DMIN(97) /* <=88 */,   DMIN(103) /* <=96 */,
+      DMIN(109) /* <=104 */, DMIN(127) /* <=112 */, DMIN(131) /* <=120 */,
+      DMIN(137) /* <=128 */, DMIN(149) /* <=136 */, DMIN(157) /* <=144 */,
+      DMIN(163) /* <=152 */, DMIN(173) /* <=160 */, DMIN(179) /* <=168 */,
+      DMIN(191) /* <=176 */, DMIN(193) /* <=184 */, DMIN(199) /* <=192 */,
+      DMIN(211) /* <=200 */, DMIN(223) /* <=208 */, DMIN(227) /* <=216 */,
+      DMIN(233) /* <=224 */, DMIN(241) /* <=232 */, DMIN(251) /* <=240 */,
+      DMIN(257) /* <=248 */, DMIN(263) /* <=256 */,
+  };
+  return MIN_PRIME + (uint32_t)kPrime[(n - 1) >> 3];
+}
+
+// Select address based on mask. mask should be all '1's or all '0's.
+// Convinience function adjusted to used types.
+static uintptr_t select_addr(uintptr_t mask, uintptr_t addr_ones,
+                             void *addr_zeroes) {
+  return (mask & addr_ones) | (~mask & (uintptr_t)addr_zeroes);
+}
+
+// Memory copy with shuffled memory access pattern
+hardened_bool_t shuffled_memcpy(void *restrict dest, const void *restrict src,
+                                size_t len, uint32_t seed) {
+  uintptr_t dest_addr = (uintptr_t)dest;
+  uintptr_t src_addr = (uintptr_t)src;
+  uintptr_t tail = dest_addr + len;
+  // Use volatile to reload tail_copy and not reduce math
```

Comment:
I think it's worth specifying which math you're trying to supress. I think this would be better served by a value barrier (I have a document that I'm still working on, but we should probably just leave a comment here and come back to it later.)

For context, the value barrier will be a function like `uint32_t launder_value(uint32_t);` that takes a value, destroys the compiler's knowledge of it, and returns it. This would allow you to prevent strength reduction as with volatile, without spilling to the stack or having other unpleasant effects, while allowing for certain reorderings that don't affect the hardening to occur.

Reply:
good point. will add launder() implementation

Reply:
I'm actually going to add this in a separate CL (and it definitely should be, because of its delicate nature). Please rebase and use that once I've done it.